### PR TITLE
CSI raw block volume support

### PR DESCRIPTION
**Pull request by BaiHuoYu**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
CSI raw block volume support
**Special notes for your reviewer**:

**Release note**:

```release-note
CSI raw block volume support
```

### CODE REVIEW

File: csi/server/plugin/opensds/node.go [hunk 1]
```diff
@@ -304,25 +358,35 @@ func (p *Plugin) NodeUnstageVolume(
...
 		return nil, err
 	}
 
-	//check volume is unmounted
-	mounted, err := connector.IsMounted(req.StagingTargetPath)
-	if !mounted {
-		glog.Info("target path is already unmounted")
-		return &csi.NodeUnstageVolumeResponse{}, nil
+	if KCSIFilesystem == vol.Metadata[KCSIVolumeMode] {
+		//check volume is unmounted
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: Add space before "check".

File: csi/server/examples/kubernetes/CreateVolumefromSnapshot/pvc_sc.yaml [hunk 2]
```diff
@@ -7,7 +7,7 @@ metadata:
   name: csi-sc-opensdsplugin-1
 provisioner: csi-opensdsplugin
 parameters:
-
+  fsType: ext4
```
Comment (xing-yang) [inline csi/server/examples/kubernetes/CreateVolumefromSnapshot/pvc_sc.yaml]: Is this needed for raw block mode?

File: csi/server/examples/kubernetes/nginx.yaml [hunk 3]
```diff
@@ -7,7 +7,7 @@ metadata:
   name: csi-sc-opensdsplugin
 provisioner: csi-opensdsplugin
 parameters:
-
+  fsType: ext4
```
Comment (xing-yang) [inline csi/server/examples/kubernetes/nginx.yaml]: Is this needed for raw block mode?

File: csi/server/examples/kubernetes/CreateVolumefromSnapshot/pvc_sc.yaml (hunk 2 above)
Comment (BaiHuoYu) [reply csi/server/examples/kubernetes/CreateVolumefromSnapshot/pvc_sc.yaml]: This has nothing to do with the raw block, just to ensure that the basic example is successful.

File: csi/server/examples/kubernetes/nginx.yaml (hunk 3 above)
Comment (BaiHuoYu) [reply csi/server/examples/kubernetes/nginx.yaml]: This has nothing to do with the raw block, just to ensure that the basic example is successful.

File: csi/server/plugin/opensds/node.go [hunk 4]
```diff
@@ -232,56 +261,81 @@ func (p *Plugin) NodeStageVolume(
...
+			}
+		} else {
+			glog.Infof("Device: %s has been formatted yet. fsType: %s", device, curFSType)
+		}
 
-	glog.Info("mounting...")
+		// Mount
+		mounted, err := connector.IsMounted(mountpoint)
+		if err != nil {
+			msg := fmt.Sprintf("Failed to check mounted, %v", err)
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/node.go]: I think it should be lower case

File: csi/server/plugin/opensds/node.go [hunk 5]
```diff
@@ -232,56 +261,81 @@ func (p *Plugin) NodeStageVolume(
...
+		if mounted {
+			glog.Info("volume is already mounted.")
+			return &csi.NodeStageVolumeResponse{}, nil
+		}
+
+		glog.Info("mounting...")
+
+		err = connector.Mount(device, mountpoint, hopeFSType, mnt.MountFlags)
+		if err != nil {
+			msg := fmt.Sprintf("Failed to mount, %v", err)
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/node.go]: Same here

File: csi/server/plugin/opensds/node.go [hunk 6]
```diff
@@ -232,56 +261,81 @@ func (p *Plugin) NodeStageVolume(
...
+			glog.Errorf(msg)
+			return nil, status.Errorf(codes.FailedPrecondition, "%s", msg)
+		}
+	} else {
+		vol.Metadata[KCSIVolumeMode] = "Block"
+		err = createSymlink(device, mountpoint)
+
+		if err != nil {
+			glog.Errorf("Failed to create a link: oldname=%v, newname=%v\n", device, mountpoint)
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/node.go]: Same here

File: csi/server/plugin/opensds/node.go [hunk 7]
```diff
@@ -375,41 +439,55 @@ func (p *Plugin) NodePublishVolume(
...
+		if err != nil {
+			msg := fmt.Sprintf("Failed to mount, %v", err)
+			glog.Errorf(msg)
+			return nil, status.Errorf(codes.FailedPrecondition, "%s", msg)
+		}
+	} else {
+		err = createSymlink(device, mountpoint)
+
+		if err != nil {
+			glog.Errorf("Failed to create a link: oldname=%v, newname=%v\n", device, mountpoint)
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/node.go]: Same here

File: csi/server/plugin/opensds/node.go [hunk 8]
```diff
@@ -440,25 +518,34 @@ func (p *Plugin) NodeUnpublishVolume(
...
+		if !mounted {
+			glog.Info("target path is already unmounted")
+			return &csi.NodeUnpublishVolumeResponse{}, nil
+		}
+
+		// Umount
+		glog.Infof("[NodeUnpublishVolume] mountpoint:%s", req.TargetPath)
+		err = connector.Umount(req.TargetPath)
+		if err != nil {
+			msg := fmt.Sprintf("Failed to Umount, %v", err)
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/node.go]: Also here

File: csi/server/plugin/opensds/node.go [hunk 9]
```diff
@@ -168,6 +170,33 @@ func delTargetPathInAttachment(attachment *model.VolumeAttachmentSpec, key strin
 	return nil
 }
 
+// Symlink implementation
+func createSymlink(device, mountpoint string) error {
+	_, err := os.Lstat(mountpoint)
+
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/node.go]: Please remove the blank line here

File: csi/server/plugin/opensds/node.go [hunk 10]
```diff
@@ -168,6 +170,33 @@ func delTargetPathInAttachment(attachment *model.VolumeAttachmentSpec, key strin
...
+		if nil != err {
+			glog.Errorf("Faild to delete %v", mountpoint)
+			return err
+		}
+	}
+
+	err = os.Symlink(device, mountpoint)
+	if err != nil {
+		glog.Errorf("Failed to create a link: oldname=%v, newname=%v\n", device, mountpoint)
```
Comment (leonwanghui) [inline csi/server/plugin/opensds/node.go]: All error msg should start from lower case

**Pull request merged**